docs/DifferentiableProgrammingImplementation.md

```diff
@@ -0,0 +1,1012 @@
+  %14 = tuple ()
+  // function_ref thunk for @escaping @callee_guaranteed (@in_guaranteed Float) -> (@out Float)
+  %15 = function_ref @$sS2fIegnr_S2fIegyd_TR : $@convention(thin) (Float, @guaranteed @callee_guaranteed (@in_guaranteed Float) -> @out Float) -> Float // user: %16
+  %16 = partial_apply [callee_guaranteed] %15(%12) : $@convention(thin) (Float, @guaranteed @callee_guaranteed (@in_guaranteed Float) -> @out Float) -> Float // user: %20
+  %17 = load [trivial] %4 : $*Float               // user: %23
+  dealloc_stack %4 : $*Float                      // id: %18
+  dealloc_stack %2 : $*Float                      // id: %19
+  %20 = struct $_AD__$sS2fIegnr_S2fIegyd_TR_bb0__PB__src_0_wrt_0 (%16 : $@callee_guaranteed (Float) -> Float) // user: %22
+  // function_ref AD__$sS2fIegnr_S2fIegyd_TR__pullback_src_0_wrt_0
+  %21 = function_ref @AD__$sS2fIegnr_S2fIegyd_TR__pullback_src_0_wrt_0 : $@convention(thin) (Float, @owned _AD__$sS2fIegnr_S2fIegyd_TR_bb0__PB__src_0_wrt_0) -> Float // user: %22
+  %22 = partial_apply [callee_guaranteed] %21(%20) : $@convention(thin) (Float, @owned _AD__$sS2fIegnr_S2fIegyd_TR_bb0__PB__src_0_wrt_0) -> Float // user: %23
+  %23 = tuple (%17 : $Float, %22 : $@callee_guaranteed (Float) -> Float) // user: %24
+  return %23 : $(Float, @callee_guaranteed (Float) -> Float) // id: %24
+} // end sil function 'AD__$sS2fIegnr_S2fIegyd_TR__vjp_src_0_wrt_0'
+```
+
+- The reabstraction thunk VJP contains `%6 = differentiable_function [parameters 0] %5`, which must now be canonicalized. This starts by finding the underlying referenced original function.
+- 💥`%5` is a function argument, so there is no underlying referenced original function! Differentiation fails and produces a non-differentiability error:
+
+```
+<unknown>:0: error: expression is not differentiable
+<unknown>:0: note: opaque non-'@differentiable' function is not differentiable
+```
+
+To support reabstraction thunk generation, we must find a way to avoid this “opaque non-@differentiable function” error. One straightforward solution is to change when the `@differentiable` function is formed:
+- Make the reabstraction thunk JVP/VJP take a `@differentiable` function-typed argument.
+- Make reabstraction thunk JVP/VJP callers construct and pass a `@differentiable` function-typed value.
+
+This involves:
+- Adding a JVP/VJP type calculation special case for reabstraction thunks.
+- Changing the differentiation transform so that reabstraction thunk JVP/VJP callers always construct and pass a `@differentiable` function-typed value.
+
+This approach is implemented in https://github.com/apple/swift/pull/28570. A partially-applied reabstraction thunk derivative matches the derivative type of the reabstracted original function.
+
+Alternatives:
+- Make reabstraction thunk JVPs/VJPs take a “JVP/VJP” function-typed argument instead of a `@differentiable` function-typed argument.
```

Comment:
```suggestion
- Make reabstraction thunk JVPs/VJPs take a "JVP/VJP" function-typed argument instead of a `@differentiable` function-typed argument.
```